File: src/Programmers/Lv3/Q2/Q2.py
"""
# Author    : Lee In Bok 
# Date      : 2023.12.31(Sun)
# Problem   : 거스름돈
# Spend Time: 15m 01s
# Runtime   : 457.90 ms (Longest Time)
# Memory    : 13 MB (Most Spacious)
# Algoritm  : Dynamic Programming
"""

def solution(n, money):
    dp = [0 for _ in range(n + 1)]
    dp[0] = 1

    for i in money:
        for j in range(i, n + 1):
            dp[j] += dp[j - i]

    return dp[n]

res = solution(5, [1,2,5])
print(res)

File: src/Programmers/Lv3/Q2/Q2.java
/**
* Author    : Lee In Bok 
* Date      : 2023.12.31(Sun)
* Problem   : 거스름돈
* Spend Time: 15m 01s
* Runtime   : 16.95 ms (Longest Time)
* Memory    : 80.8 MB (Most Spacious)
* Algoritm  : Dynamic Programming
 */

package Programmers.Lv3.Q2;

public class Q2 {
  public static void main(String[] args) {
    Solution s = new Solution();
    int res = s.solution(5, new int[]{1, 2, 5});

    System.out.println(res);
  }
}

class Solution {
  public int solution(int n, int[] money) {
    int[] dp = new int[n + 1];
    dp[0] = 1;

    for(int i = 0; i < money.length; i++){
      for(int j = money[i]; j < n + 1; j++){
        dp[j] += dp[j - money[i]];
      }
    }

    return dp[n];
  }
}